File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        val dummy = ListNode(0)
        var cur = dummy
        var l1 = list1
        var l2 = list2
        while (l1 !== null && l2 !== null) {
            if (l1.`val` < l2.`val`) {
                cur.next = l1
                l1 = l1.next
            } else {
                cur.next = l2
                l2 = l2.next
            }
            cur = cur.next
        }
        if (l1 !== null) {
            cur.next = l1
        }
        if (l2 !== null) {
            cur.next = l2
        }
        return dummy.next
    }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    if (!list1) return list2
    if (!list2) return list1
    const cur = list1.val < list2.val ? list1 : list2
    const notCur = cur === list1 ? list2 : list1
    cur.next = mergeTwoLists(cur.next, notCur)
    return cur
};